src/GovUk.Education.ExploreEducationStatistics.Common/Rules/LowercasePathRule.cs

```diff
@@ -0,0 +1,45 @@
+﻿#nullable enable
+using System.Linq;
+using System.Text.RegularExpressions;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Rewrite;
+using Microsoft.Net.Http.Headers;
+
+namespace GovUk.Education.ExploreEducationStatistics.Common.Rules;
+
+public class LowercasePathRule : IRule
+{
+    private const string ExcludedPathsRegex = "^/(_configuration|identity|static).*$";
+
+    public void ApplyRule(RewriteContext context)
+    {
+        var request = context.HttpContext.Request;
+
+        var scheme = request.Scheme;
+        var pathBase = request.PathBase;
+        var path = request.Path;
+        var host = request.Host;
+        var query = request.QueryString;
+
+        if (IsApplicable(path))
+        {
+            var location = $"{scheme}://{host.Value}{pathBase.Value}{path.Value}".ToLower() + query;
```

Comment:
Just wondering, but did we consider just trimming the spacing here? Feels like it'd be simpler than adding all the additional model binding code?

Reply:
Yes I did consider this but trimming `path.Value` here wouldn't trim whitespace from parameters where a route template contains the parameters mid-way through it rather than just at the end, e.g. `/api/publications/{publication-slug}/releases/latest` or `/api/publications/{publication-slug}/releases/{release-slug}`.

Did you mean trying to remove whitespace part way through the path string? I didn't think that would be safe as I couldn't think of a way of just targeting the route parameters here without knowing them in advance to trim them.

The model binder solution for trimming comes with the bonus of trimming parameters that come from value providers for query string values and form values as well as route values (see `TrimStringModelBinderIntegrationTests` for examples).

Aside from the problem with caching I'm trying to solve here, I really wanted to implement a model binder that would trim body values too for complex types using `[FromBody]`. Due to the way values are read from the message body using a JSON media-type formatter and deserialised, a one-fits-all solution to trimming wasn't possible so I gave up with that.

Reply:
Couldn't we just strip out any whitespace with a regex? Spaces are not actually allowed in URLs at all, so feels like we should be able to do this.

In fact, a similar problem exists with the model binding implementation as we're not accounting for permutations of things where the whitespace is within the path segment e.g. `some-path fragment`

Reply:
I agree URL's should not contain spaces but spaces can exist in URL's in URL encoded notation (decoded in the request prior to model binding) and I think it's safer to only target parameters that are being bound to the model and sanitise those by trimming leading and trailing whitespace from them instead of removing spaces from a URL indiscriminately.

The model binder implemented here never intends to remove whitespace from _within_ parameters,  e.g. your example `some-path segment` would be left alone, just as I would expect it to be after sanitisation if it's a string parameter bound to the model.

Also spaces _within_ parameters that are used for cache keys don't cause the problem with caching that I'm trying to address here. Parameters which go on to be used as keys in a database lookup after a cache-miss will find a result only if there's an exact match just as we'd expect, unlike if they contain trailing whitespace where they still find a result.

It could be argued that spaces within route parameters or other parts of the URL which shouldn't be there (e.g. introduced in to parts of the URL by something other than not sanitising user inputs from leading/trailing whitespace) should be fixed at source and return 404.

I agree we could make the change you suggest here, changing the rule to remove all spaces as well as lowercasing like it does now and that it would have the advantage of returning a clean redirect the client.

I guess we'd have to tread carefully around removing spaces from the query string, which could genuinely contain spaces, so we'd only want to remove all spaces before the query string, and then either leave the query string untouched (same as with the lowercasing), or go on to just trim individual query string parameters without removing spaces indiscriminately. As I mentioned before, the model binder implemented here will already handle trimming parameters from query sources and form sources too.

Apart from the redirect I'm not convinced that there's any benefit to doing this with a rule over what I've already implemented with the model binder, or that it's worth the effort changing it.

Reply:
If we did make this change here, you're right that it'd be best to only do it on the main part of the URL and not the query string.

If we're worried about it being a bit too aggressive of a measure, happy for us not to do this. As long as we get the intended behaviour for malformed URLs (404s) instead of a cached result, it's fine. From what you're saying, this seems to be the case, so happy to go with it.

I was thinking we'd be able to simplify the code a little bit without the model binders, but as you've said, the model binders also trim query and form input. That alone seems worth it to keep them in!

Reply:
Ok thanks for your time spent considering the options here. It's been interesting understanding how controller actions and model binding work and weighing up the different options for doing this.

If it turns out to not work or not do what we want then we can switch plan! 🙂